lib/Sema/TypeCheckType.cpp

```diff
@@ -2256,6 +2256,15 @@ Type TypeResolver::resolveAttributedType(TypeAttributes &attrs,
                    attrs.getConvention());
           rep = FunctionType::Representation::Swift;
         } else {
+          auto convention = attrs.getConvention();
+          if (attrs.has(TAK_autoclosure)) {
+            // @convention(c) and @convention(block) are not allowed with an @autoclosure type.
+            if (convention == "c" || convention == "block" ) {
```

Comment:
We already check for `FunctionTypeRepresentation` above (see `rep`), maybe you can use it here? You will need to compare it with `FunctionTypeRepresentation::CFunctionPointer` and `FunctionTypeRepresentation::Block`. 

You can even move the check down inside `if (!isParam && attrs.has(TAK_autoclosure)) { ... }` block, because we already know we have an `@autoclosure` there.